File: Samples/HelloTriangle.ANGLE/SampleForm.cs

// Copyright (C) 2016-2017 Luca Piccioni
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

using System;
using System.Text;
using System.Windows.Forms;

using OpenGL;

namespace HelloTriangle.ANGLE
{
	/// <summary>
	/// Hello triangle sample, using ANGLE for providing OpenGL ES on Windows OS.
	/// </summary>
	public partial class SampleForm : Form
	{
		/// <summary>
		/// Construct a SampleForm.
		/// </summary>
		public SampleForm()
		{
			InitializeComponent();
		}

		private void SampleForm_Load(object sender, EventArgs e)
		{
			
		}

		private void RenderControl_ContextCreated(object sender, GlControlEventArgs e)
		{
			// Update Form caption
			Text = String.Format("Hello triangle with ANGLE (Version: {0})", Gl.GetString(StringName.Version));

			// Create resources
			StringBuilder infolog = new StringBuilder(1024);
			int infologLength;
			int compiled;

			infolog.EnsureCapacity(1024);

			// Vertex shader
			uint vertexShader = Gl.CreateShader(ShaderType.VertexShader);
			Gl.ShaderSource(vertexShader, _Es2_ShaderVertexSource);
			Gl.CompileShader(vertexShader);
			Gl.GetShader(vertexShader, ShaderParameterName.CompileStatus, out compiled);
			if (compiled == 0) {
				Gl.GetShaderInfoLog(vertexShader, 1024, out infologLength, infolog);
			}

			// Fragment shader
			uint fragmentShader = Gl.CreateShader(ShaderType.FragmentShader);
			Gl.ShaderSource(fragmentShader, _Es2_ShaderFragmentSource);
			Gl.CompileShader(fragmentShader);
			Gl.GetShader(fragmentShader, ShaderParameterName.CompileStatus, out compiled);
			if (compiled == 0) {
				Gl.GetShaderInfoLog(fragmentShader, 1024, out infologLength, infolog);
			}

			// Program
			_Es2_Program = Gl.CreateProgram();
			Gl.AttachShader(_Es2_Program, vertexShader);
			Gl.AttachShader(_Es2_Program, fragmentShader);
			Gl.LinkProgram(_Es2_Program);

			int linked;
			Gl.GetProgram(_Es2_Program, ProgramProperty.LinkStatus, out linked);

			if (linked == 0) {
				Gl.GetProgramInfoLog(_Es2_Program, 1024, out infologLength, infolog);
			}

			_Es2_Program_Location_uMVP = Gl.GetUniformLocation(_Es2_Program, "uMVP");
			_Es2_Program_Location_aPosition = Gl.GetAttribLocation(_Es2_Program, "aPosition");
			_Es2_Program_Location_aColor = Gl.GetAttribLocation(_Es2_Program, "aColor");
		}

		private void RenderControl_Render(object sender, GlControlEventArgs e)
		{
			Control control = (Control)sender;

			Gl.Viewport(0, 0, control.Width, control.Height);
			Gl.Clear(ClearBufferMask.ColorBufferBit);

			Gl.UseProgram(_Es2_Program);

			using (MemoryLock arrayPosition = new MemoryLock(_ArrayPosition))
			using (MemoryLock arrayColor = new MemoryLock(_ArrayColor))
			{
				Gl.VertexAttribPointer((uint)_Es2_Program_Location_aPosition, 2, VertexAttribType.Float, false, 0, arrayPosition.Address);
				Gl.EnableVertexAttribArray((uint)_Es2_Program_Location_aPosition);

				Gl.VertexAttribPointer((uint)_Es2_Program_Location_aColor, 3, VertexAttribType.Float, false, 0, arrayColor.Address);
				Gl.EnableVertexAttribArray((uint)_Es2_Program_Location_aColor);

				Gl.UniformMatrix4f(_Es2_Program_Location_uMVP, 1, false, Matrix4x4f.Ortho2D(0.0f, 1.0f, 0.0f, 1.0f));

				Gl.DrawArrays(PrimitiveType.Triangles,  0, 3);
			}
		}

		private void RenderControl_ContextDestroying(object sender, OpenGL.GlControlEventArgs e)
		{
			if (_Es2_Program != 0)
				Gl.DeleteProgram(_Es2_Program);
			_Es2_Program = 0;
		}

		private float _Angle;

		private uint _Es2_Program;

		private int _Es2_Program_Location_aPosition;

		private int _Es2_Program_Location_aColor;

		private int _Es2_Program_Location_uMVP;

		private readonly string[] _Es2_ShaderVertexSource = new string[] {
			"uniform mat4 uMVP;\n",
			"attribute vec2 aPosition;\n",
			"attribute vec3 aColor;\n",
			"varying vec3 vColor;\n",
			"void main() {\n",
			"	gl_Position = uMVP * vec4(aPosition, 0.0, 1.0);\n",
			"	vColor = aColor;\n",
			"}\n"
		};

		private readonly string[] _Es2_ShaderFragmentSource = new string[] {
			"precision mediump float;\n",
			"varying vec3 vColor;\n",
			"void main() {\n",
			"	gl_FragColor = vec4(vColor, 1.0);\n",
			"}\n"
		};

		/// <summary>
		/// Vertex position array.
		/// </summary>
		private static readonly float[] _ArrayPosition = new float[] {
			0.0f, 0.0f,
			0.5f, 1.0f,
			1.0f, 0.0f
		};

		/// <summary>
		/// Vertex color array.
		/// </summary>
		private static readonly float[] _ArrayColor = new float[] {
			1.0f, 0.0f, 0.0f,
			0.0f, 1.0f, 0.0f,
			0.0f, 0.0f, 1.0f
		};
	}
}

File: Samples/HelloTriangle.VB/SampleForm.vb

' Copyright (C) 2016-2017 Luca Piccioni
' 
' Permission Is hereby granted, free of charge, to any person obtaining a copy
' of this software And associated documentation files (the "Software"), to deal
' in the Software without restriction, including without limitation the rights
' to use, copy, modify, merge, publish, distribute, sublicense, And/Or sell
' copies of the Software, And to permit persons to whom the Software Is
' furnished to do so, subject to the following conditions:
' 
' The above copyright notice And this permission notice shall be included in all
' copies Or substantial portions of the Software.
' 
' THE SOFTWARE Is PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS Or
' IMPLIED, INCLUDING BUT Not LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
' FITNESS FOR A PARTICULAR PURPOSE And NONINFRINGEMENT. IN NO EVENT SHALL THE
' AUTHORS Or COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES Or OTHER
' LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT Or OTHERWISE, ARISING FROM,
' OUT OF Or IN CONNECTION WITH THE SOFTWARE Or THE USE Or OTHER DEALINGS IN THE
' SOFTWARE.

Imports OpenGL

Public Class SampleForm

	Private Sub RenderControl_ContextCreated(sender As Object, e As OpenGL.GlControlEventArgs) Handles RenderControl.ContextCreated

		' Here you can allocate resources Or initialize state
		Gl.MatrixMode(MatrixMode.Projection)
		Gl.LoadIdentity()
		Gl.Ortho(0.0, 1.0F, 0.0, 1.0, 0.0, 1.0)

		Gl.MatrixMode(MatrixMode.Modelview)
		Gl.LoadIdentity()

	End Sub

	Private Sub RenderControl_Render(sender As Object, e As GlControlEventArgs) Handles RenderControl.Render

		Dim senderControl As Control = CType(sender, Control)

		' Note: conflicting commands are scoped in Gl.VB class
		Gl.VB.Viewport(0, 0, senderControl.ClientSize.Width, senderControl.ClientSize.Height)
		Gl.VB.Clear(ClearBufferMask.ColorBufferBit)

		' Note: conflicting enumerants are scoped in Gl.VBEnum class
		Dim currentViewport As Integer() = New Integer(3) {0, 0, 0, 0}
		Gl.Get(Gl.VBEnum.VIEWPORT, currentViewport)

		If (Gl.CurrentVersion >= Gl.Version_110) Then
			' Old school OpenGL 1.1
			' Setup & enable client states to specify vertex arrays, And use Gl.DrawArrays instead of Gl.Begin/End paradigm
			Using vertexArrayLock As New MemoryLock(_ArrayPosition)
				Using vertexColorLock As New MemoryLock(_ArrayColor)
					' Note: the use of MemoryLock objects Is necessary to pin vertex arrays since they can be reallocated by GC
					' at any time between the Gl.VertexPointer execution And the Gl.DrawArrays execution

					Gl.VertexPointer(2, VertexPointerType.Float, 0, vertexArrayLock.Address)
					Gl.EnableClientState(EnableCap.VertexArray)

					Gl.ColorPointer(3, ColorPointerType.Float, 0, vertexColorLock.Address)
					Gl.EnableClientState(EnableCap.ColorArray)

					Gl.DrawArrays(PrimitiveType.Triangles, 0, 3)
				End Using
			End Using
		Else
			' Old school OpenGL
			Gl.Begin(PrimitiveType.Triangles)
			Gl.Color3(1.0F, 0.0F, 0.0F)
			Gl.Vertex2(0.0F, 0.0F)
			Gl.Color3(0.0F, 1.0F, 0.0F)
			Gl.Vertex2(0.5F, 1.0F)
			Gl.Color3(0.0F, 0.0F, 1.0F)
			Gl.Vertex2(1.0F, 0.0F)
			Gl.End()
		End If

	End Sub

	Private Sub RenderControl_ContextDestroying(sender As Object, e As OpenGL.GlControlEventArgs) Handles RenderControl.ContextDestroying
		' Here you can dispose resources allocated in RenderControl_ContextCreated
	End Sub

	''' <summary>
	''' Vertex position array.
	''' </summary>
	Private ReadOnly _ArrayPosition As Single() = {
		0.0F, 0.0F,
		0.5F, 1.0F,
		1.0F, 0.0F
	}

	''' <summary>
	''' Vertex color array.
	''' </summary>
	Private ReadOnly _ArrayColor As Single() = {
		1.0F, 0.0F, 0.0F,
		0.0F, 1.0F, 0.0F,
		0.0F, 0.0F, 1.0F
	}
End Class
